### PR TITLE
Subs 491 ena data model validation

### PR DESCRIPTION
**Pull request by neilgoodgame**
Uses three new Annotations for validations 

ENAFieldAttribute for validating that an attribute is present
ENAControlledValueAttribute for validating that an attribute is one of an allowed controlled value
ENAField for copying an attribute to a field in the ENASubmittable.  This can also contain a controlled value.  This is used for the platforms in the ENAExperiment.

### CODE REVIEW

File: subs-ena-data-model/src/main/java/uk/ac/ebi/subs/ena/validation/AttributeRequiredValidationResult.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import uk.ac.ebi.ena.sra.xml.RECEIPTDocument;
+import uk.ac.ebi.subs.data.submittable.ENASubmittable;
+import uk.ac.ebi.subs.validator.data.ValidationStatus;
+
+/**
+ * Created by neilg on 14/06/2017.
+ */
+public class AttributeRequiredValidationResult extends AbstractENAValidationResult {
+    private static String MESSAGE = "Value for attribute %s is required.";
+    String attributeName;
```
Comment (ke4) [inline subs-ena-data-model/src/main/java/uk/ac/ebi/subs/ena/validation/AttributeRequiredValidationResult.java]: Do we need this variable?

File: subs-ena-data-model/src/main/java/uk/ac/ebi/subs/ena/validation/InvalidAttributeValue.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import org.apache.commons.lang.StringUtils;
+import uk.ac.ebi.subs.data.submittable.ENASubmittable;
+
+/**
+ * Created by neilg on 14/06/2017.
+ */
+public class InvalidAttributeValue extends AbstractENAValidationResult {
+    static final String INVALID_VALUE_ERROR_MESSAGE = "Invalid value %s for attribute %s value must be one of %s.";
+
+    String attributeValue;
```
Comment (ke4) [inline subs-ena-data-model/src/main/java/uk/ac/ebi/subs/ena/validation/InvalidAttributeValue.java]: You never set this value and you probably don't need this

File: subs-ena-data-model/src/main/java/uk/ac/ebi/subs/ena/validation/InvalidAttributeValue.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    String[] allowedValues;
```
Comment (ke4) [inline subs-ena-data-model/src/main/java/uk/ac/ebi/subs/ena/validation/InvalidAttributeValue.java]: The same as above

File: subs-ena-data-model/src/main/java/uk/ac/ebi/subs/ena/validation/Validator.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+import uk.ac.ebi.subs.validator.data.SingleValidationResult;
+
+import java.lang.reflect.Field;
+import java.util.List;
+
+/**
+ * Created by neilg on 04/07/2017.
+ */
+public interface Validator  {
```
Comment (ke4) [inline subs-ena-data-model/src/main/java/uk/ac/ebi/subs/ena/validation/Validator.java]: Where do you use this interface?
I can't see it anywhere, but I might be wrong

File: subs-ena-data-model/src/test/java/uk/ac/ebi/subs/xml/ExperimentSerialisationTest.java [hunk 5]
```diff
@@ -217,6 +223,11 @@ public void testExperimentHelicosPlatform() throws Exception {
         testPlatform(HELICOS_INSTRUMENT_MODEL_XPATH,HELICOS, HELICOS_HELISCOPE_INSTRUMENT_MODEL);
     }
 
+    //@Test
```
Comment (ke4) [inline subs-ena-data-model/src/test/java/uk/ac/ebi/subs/xml/ExperimentSerialisationTest.java]: Is it a typo to leave this test out?

File: subs-ena-data-model/src/test/java/uk/ac/ebi/subs/xml/ExperimentSerialisationTest.java [hunk 6]
```diff
@@ -261,15 +272,34 @@ public void testPlatform(String platformXpathQuery, String plaformType, String i
         assertThat("experiment alias serialised to XML", instrumentModel, equalTo(returnedInstrumentModel));
     }
 
-    @Test(expected = IllegalArgumentException.class)
+    public void testInvalidPlatform(String platformXpathQuery, String plaformType) throws Exception {
```
Comment (ke4) [inline subs-ena-data-model/src/test/java/uk/ac/ebi/subs/xml/ExperimentSerialisationTest.java]: Should not it be a private method?

File: subs-ena-data-model/src/test/java/uk/ac/ebi/subs/xml/ExperimentSerialisationTest.java [hunk 7]
```diff
@@ -261,15 +272,34 @@ public void testPlatform(String platformXpathQuery, String plaformType, String i
...
+            }
+        }
+        assertThat(invalidAttributeValue, is(IsNull.notNullValue()));
+
         final Document document = documentBuilderFactory.newDocumentBuilder().newDocument();
         marshaller.marshal(enaExperiment,new DOMResult(document));
     }
 
-    @Test(expected = IllegalArgumentException.class)
+    //@Test(expected = IllegalArgumentException.class)
```
Comment (ke4) [inline subs-ena-data-model/src/test/java/uk/ac/ebi/subs/xml/ExperimentSerialisationTest.java]: Why is it commented out?

File: subs-ena-data-model/src/test/java/uk/ac/ebi/subs/xml/ExperimentSerialisationTest.java [hunk 8]
```diff
@@ -345,7 +375,7 @@ protected ENASubmittable createENASubmittable() throws IllegalAccessException {
         return new ENAExperiment();
     }
 
-    @Test
+    //@Test
```
Comment (ke4) [inline subs-ena-data-model/src/test/java/uk/ac/ebi/subs/xml/ExperimentSerialisationTest.java]: Same question as above

File: subs-ena-data-model/src/test/java/uk/ac/ebi/subs/xml/StudySerialisationTest.java [hunk 9]
```diff
@@ -142,6 +146,31 @@ public void testMarshalStudyType() throws Exception {
...
+        String incorrectStudyType = UUID.randomUUID().toString();
+        attribute.setValue(incorrectStudyType);
+        study.getAttributes().add(attribute);
+        ENAStudy enaStudy = new ENAStudy(study);
+        final Document document = documentBuilderFactory.newDocumentBuilder().newDocument();
+        marshaller.marshal(enaStudy,new DOMResult(document));
+        final List<SingleValidationResult> validationResultList = enaStudy.getValidationResultList();
+
+        assertThat("Study is invalid",enaStudy.isValid(),equalTo(false));
+        /*
```
Comment (ke4) [inline subs-ena-data-model/src/test/java/uk/ac/ebi/subs/xml/StudySerialisationTest.java]: you can remove these commented lines

File: subs-ena-data-model/src/main/java/uk/ac/ebi/subs/data/submittable/ENAExperiment.java [hunk 10]
```diff
@@ -30,63 +44,64 @@
...
             "AB SOLiD 4 System", "AB SOLiD 4hq System", "AB SOLiD PI System", "AB 5500 Genetic Analyzer", "AB 5500xl Genetic Analyzer",
             "AB 5500xl-W Genetic Analysis System", "unspecified"})
     String abiSolid ;
 
-    @ENAPlatform(name = "COMPLETE_GENOMICS", instrumentModels = {"Complete Genomics", "unspecified"})
+    @ENAField(fieldName = "COMPLETE_GENOMICS", values = {"Complete Genomics", "unspecified"})
     String completeGenomics;
 
-    @ENAPlatform(name = "BGISEQ", instrumentModels = {"BGISEQ-500"})
+    @ENAField(fieldName = "BGISEQ", values = {"BGISEQ-500"})
```
Comment (neilgoodgame) [inline subs-ena-data-model/src/main/java/uk/ac/ebi/subs/data/submittable/ENAExperiment.java]: Not for platform BGISEQ, defined in the XSD as 

<xs:simpleType name="typeBGISEQModel">
        <xs:restriction base="xs:string">
            <xs:enumeration value="BGISEQ-500"/>
        </xs:restriction>
    </xs:simpleType>

File: subs-ena-data-model/src/main/java/uk/ac/ebi/subs/ena/validation/Validator.java (hunk 4 above)
Comment (neilgoodgame) [inline subs-ena-data-model/src/main/java/uk/ac/ebi/subs/ena/validation/Validator.java]: Removed it

File: subs-ena-data-model/src/test/java/uk/ac/ebi/subs/xml/ExperimentSerialisationTest.java (hunk 6 above)
Comment (neilgoodgame) [inline subs-ena-data-model/src/test/java/uk/ac/ebi/subs/xml/ExperimentSerialisationTest.java]: changed to private

File: subs-ena-data-model/src/main/java/uk/ac/ebi/subs/data/submittable/ENAExperiment.java (hunk 10 above)
Comment (ke4) [inline subs-ena-data-model/src/main/java/uk/ac/ebi/subs/data/submittable/ENAExperiment.java]: OK

File: subs-ena-data-model/build.gradle [hunk 11]
```diff
@@ -2,13 +2,17 @@ apply plugin: 'java'
...
-    compile("uk.ac.ebi.subs:subs-data-model:1.0.3-SNAPSHOT")
-
+    compile("uk.ac.ebi.subs:subs-data-model:1.0.1-SNAPSHOT")
     compile ("org.springframework:spring-oxm")
     compile 'org.eclipse.persistence:org.eclipse.persistence.moxy:2.6.4'
     compile 'commons-lang:commons-lang:2.6'
     testCompile("org.springframework.boot:spring-boot-starter-test")
     testCompile project(':subs-test-categories')
+    compile 'uk.ac.ebi.subs:subs-data-model:1.0.1-SNAPSHOT'
+    compile ("uk.ac.ebi.subs.validator.common:validator-common:1.0.1-SNAPSHOT") {
```
Comment (ke4) [inline subs-ena-data-model/build.gradle]: it should be: `compile ("uk.ac.ebi.subs:validator-common:1.11.0-SNAPSHOT")`

**Comment by neilgoodgame** [comment]
I've fixed the deserialisation for platform type and instrument model

**Pull request merged**